# Coding Exercises (Part 2)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas Series an Index Objects


### Exercise 3: Analyzing Columns / Pandas Series

In [1]:
#run the cell!
import pandas as pd

In [2]:
#run the cell!
cars = pd.read_csv("cars.csv")

In [3]:
#run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

In [4]:
#run the cell!
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
mpg             398 non-null float64
cylinders       398 non-null int64
displacement    398 non-null float64
horsepower      392 non-null float64
weight          398 non-null int64
acceleration    398 non-null float64
model_year      398 non-null int64
origin          398 non-null object
name            398 non-null object
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


18. Select the numerical column __"mpg"__, create a __copy__ and __save__ the column/Pandas Series in the variable __mpg__! __Fill in the gaps__!

In [5]:
mpg = cars.mpg.copy()

In [6]:
#run the cell!
mpg.head()

0    18.0
1    15.0
2    18.0
3    16.0
4    17.0
Name: mpg, dtype: float64

19. Get some __summary statistics__ on the Series __mpg__! What is the __mpg__ of the __least fuel efficient__ car?

In [7]:
mpg.describe()

count    398.000000
mean      23.514573
std        7.815984
min        9.000000
25%       17.500000
50%       23.000000
75%       29.000000
max       46.600000
Name: mpg, dtype: float64

The mpg of the least fuel efficient car is ... 9.0 miles per gallon!

20. Get the __maximum Value__ in the Series __mpg__ by explicitly calling the ... method! The __most fuel efficient__ car has a __mpg__ of...?

In [8]:
mpg.max()

46.6

The most fuel efficient car has a mpg of... 46.6!

21. Get the Frequency/__Counts__ of all unique __values__ in the Series __mpg__! What is the __most frequent__ value?

In [9]:
mpg.value_counts()

13.0    20
14.0    19
18.0    17
15.0    16
26.0    14
16.0    13
19.0    12
24.0    11
25.0    11
28.0    10
22.0    10
23.0    10
20.0     9
27.0     9
29.0     8
21.0     8
30.0     7
31.0     7
17.0     7
12.0     6
36.0     6
32.0     6
17.5     5
15.5     5
11.0     4
38.0     4
20.2     4
21.5     3
33.0     3
18.5     3
        ..
20.8     1
23.6     1
30.9     1
43.4     1
39.1     1
40.9     1
20.3     1
34.7     1
32.9     1
22.3     1
46.6     1
32.2     1
26.8     1
19.1     1
27.5     1
19.9     1
41.5     1
31.6     1
35.7     1
23.5     1
39.0     1
25.1     1
32.8     1
44.0     1
28.8     1
34.2     1
38.1     1
37.2     1
32.1     1
32.7     1
Name: mpg, Length: 129, dtype: int64

The most frequent value is ... 13.0!

22. Get the __relative frequencies__ in the Series __mpg__! What is the __relative frequency__ of the __most frequent value__?

In [10]:
mpg.value_counts(normalize = True)

13.0    0.050251
14.0    0.047739
18.0    0.042714
15.0    0.040201
26.0    0.035176
16.0    0.032663
19.0    0.030151
24.0    0.027638
25.0    0.027638
28.0    0.025126
22.0    0.025126
23.0    0.025126
20.0    0.022613
27.0    0.022613
29.0    0.020101
21.0    0.020101
30.0    0.017588
31.0    0.017588
17.0    0.017588
12.0    0.015075
36.0    0.015075
32.0    0.015075
17.5    0.012563
15.5    0.012563
11.0    0.010050
38.0    0.010050
20.2    0.010050
21.5    0.007538
33.0    0.007538
18.5    0.007538
          ...   
20.8    0.002513
23.6    0.002513
30.9    0.002513
43.4    0.002513
39.1    0.002513
40.9    0.002513
20.3    0.002513
34.7    0.002513
32.9    0.002513
22.3    0.002513
46.6    0.002513
32.2    0.002513
26.8    0.002513
19.1    0.002513
27.5    0.002513
19.9    0.002513
41.5    0.002513
31.6    0.002513
35.7    0.002513
23.5    0.002513
39.0    0.002513
25.1    0.002513
32.8    0.002513
44.0    0.002513
28.8    0.002513
34.2    0.002513
38.1    0.002513
37.2    0.0025

The relative frequency of 13.0 is ... 0.050251!

23. __Sort__ the Series __mpg__ from __low to high__! What is the __second lowest__ value?

In [11]:
mpg.sort_values()

28      9.0
25     10.0
26     10.0
103    11.0
124    11.0
27     11.0
67     11.0
95     12.0
104    12.0
90     12.0
106    12.0
42     12.0
69     12.0
213    13.0
85     13.0
44     13.0
43     13.0
215    13.0
87     13.0
214    13.0
62     13.0
92     13.0
73     13.0
94     13.0
96     13.0
208    13.0
137    13.0
70     13.0
68     13.0
105    13.0
       ... 
381    36.0
379    36.0
391    36.0
380    36.0
245    36.1
248    36.1
327    36.4
347    37.0
320    37.0
376    37.0
312    37.2
304    37.3
348    37.7
387    38.0
378    38.0
385    38.0
383    38.0
310    38.1
344    39.0
343    39.1
247    39.4
324    40.8
330    40.9
309    41.5
244    43.1
326    43.4
394    44.0
325    44.3
329    44.6
322    46.6
Name: mpg, Length: 398, dtype: float64

The second lowest value is ... 10.0!

24. __Sort__ the Series __mpg__ from __high to low__ and __save the changes__ by setting the __inplace__ parameter to __True__!  <br>
__Fill in the gaps!__

In [12]:
mpg.sort_values(ascending = False, inplace = True)

Inspect the __first 5 elements__ of mpg! What is the __second highest__ value?

In [13]:
#run the cell!
mpg.head()

322    46.6
329    44.6
325    44.3
394    44.0
326    43.4
Name: mpg, dtype: float64

The second highest value is ... 44.6!

25. __Sort__ the Series __mpg__ by the __Index__ and __save__ the changes!

In [14]:
mpg.sort_index(inplace = True)

In [15]:
#run the cell!
mpg.head(10)

0    18.0
1    15.0
2    18.0
3    16.0
4    17.0
5    15.0
6    14.0
7    14.0
8    14.0
9    15.0
Name: mpg, dtype: float64

__Miles per Gallon__ (mpg) can be transformed into __Liter per 100 Kilometer__ with the following formula:<br>
__Liter per 100 Kilometer = 235.21 / mpg__<br><br>
26. Create a new Pandas Series __l_per_100__ by applying the above formula! __Round__ the results to __2 decimals__! __Fill in the gaps__!

In [16]:
l_per_100 = (235.21/mpg).round(2)

Run and Inspect. What is the __very first__ element?

In [17]:
#run the cell!
l_per_100.head()

0    13.07
1    15.68
2    13.07
3    14.70
4    13.84
Name: mpg, dtype: float64

The first element is ... 13.07!

27. Get some __summary statistics__ on the Series __l_per_100__! What is the __average__ value?

In [18]:
l_per_100.describe()

count    398.000000
mean      11.212789
std        3.901407
min        5.050000
25%        8.110000
50%       10.230000
75%       13.440000
max       26.130000
Name: mpg, dtype: float64

The average value is approx. ... 11.21!

28. Select the non-numerical column __"origin"__, create a __copy__ and __save__ the column/Pandas Series in the variable __origin__! 

In [19]:
origin = cars.origin.copy()

Inspect! The first 5 elements are all...?

In [20]:
#run the cell!
origin.head()

0    usa
1    usa
2    usa
3    usa
4    usa
Name: origin, dtype: object

The first 5 elements are all ... usa!

29. Call the __describe()__ method on the non-numerical Series __"origin"__! What is the __most frequent__ value?

In [21]:
origin.describe()

count     398
unique      3
top       usa
freq      249
Name: origin, dtype: object

The most frequent value is ... usa!

30. Get all __unique values__ in the Series __origin__! Apart from the value usa, there are also the values...?

In [22]:
origin.unique()

array(['usa', 'japan', 'europe'], dtype=object)

... japan and europe!

31. Last but not least, __count the frequencies__ in the Series __origin__! How often does the value __europe__ appear?

In [23]:
origin.value_counts()

usa       249
japan      79
europe     70
Name: origin, dtype: int64

The value europe appears ... 70 times!

### Exercise 4: Index Operations

In [24]:
#run the cell!
import pandas as pd

In [25]:
#run the cell!
cars = pd.read_csv("cars.csv")

In [26]:
#run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

In [27]:
#run the cell!
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
mpg             398 non-null float64
cylinders       398 non-null int64
displacement    398 non-null float64
horsepower      392 non-null float64
weight          398 non-null int64
acceleration    398 non-null float64
model_year      398 non-null int64
origin          398 non-null object
name            398 non-null object
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


32. __Select__ the __index__ (row labels) of the DataFrame __cars__! What __kind of index__ do we have?

In [28]:
cars.index

RangeIndex(start=0, stop=398, step=1)

It´s a ... RangeIndex!

33. __Set__ the __"name"__ column as the __new index__ of cars (save the changes)!

In [29]:
cars.set_index("name", inplace = True)

34. __Check__ whether the new index contains __only unique values__ (no duplicates)! Is this the case?

In [30]:
cars.index.is_unique

False

No, apparently, there are duplicate entries!

35. __Get__ the __frequency/Counts__ of all unique values in the index! What is the __most frequent__ value?

In [31]:
cars.index.value_counts()

ford pinto                              6
ford maverick                           5
amc matador                             5
toyota corolla                          5
vw rabbit                               5
amc hornet                              4
chevrolet chevette                      4
chevrolet impala                        4
amc gremlin                             4
peugeot 504                             4
toyota corona                           4
plymouth duster                         3
chevrolet nova                          3
dodge colt                              3
chevrolet chevelle malibu               3
chevrolet citation                      3
honda civic                             3
vw dasher                               3
chevrolet vega                          3
plymouth fury iii                       3
chevrolet caprice classic               3
ford gran torino                        3
pontiac catalina                        3
ford galaxie 500                  

The most frequent value is ... ford pinto!

36. __Set__ the __name__ of the index to __"car_model"__!

In [32]:
cars.index.name = "car_model"

In [33]:
#run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  \
car_model                                                                      
chevrolet chevelle malibu  18.0          8         307.0       130.0    3504   
buick skylark 320          15.0          8         350.0       165.0    3693   
plymouth satellite         18.0          8         318.0       150.0    3436   
amc rebel sst              16.0          8         304.0       150.0    3433   
ford torino                17.0          8         302.0       140.0    3449   

                           acceleration  model_year origin  
car_model                                                   
chevrolet chevelle malibu          12.0          70    usa  
buick skylark 320                  11.5          70    usa  
plymouth satellite                 11.0          70    usa  
amc rebel sst                      12.0          70    usa  
ford torino                        10.5          70    usa

37. __Reset__ the __index__ to a RangeIndex and __save__ the change!

In [34]:
cars.reset_index(inplace= True)

In [35]:
#run the cell!
cars.head()

car_model   mpg  cylinders  displacement  horsepower  \
0  chevrolet chevelle malibu  18.0          8         307.0       130.0   
1          buick skylark 320  15.0          8         350.0       165.0   
2         plymouth satellite  18.0          8         318.0       150.0   
3              amc rebel sst  16.0          8         304.0       150.0   
4                ford torino  17.0          8         302.0       140.0   

   weight  acceleration  model_year origin  
0    3504          12.0          70    usa  
1    3693          11.5          70    usa  
2    3436          11.0          70    usa  
3    3433          12.0          70    usa  
4    3449          10.5          70    usa

38. __Inspect__ the __column Index__!

In [36]:
cars.columns

Index(['car_model', 'mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'model_year', 'origin'],
      dtype='object')

39. __Rename__ the column __Labels__ "horsepower" and "origin" to "hp" and "country"! __Fill in  the gaps__!

In [37]:
cars.rename(columns = {"horsepower":"hp", "origin":"country"}, inplace = True)

In [38]:
#run the cell!
cars.head()

car_model   mpg  cylinders  displacement     hp  weight  \
0  chevrolet chevelle malibu  18.0          8         307.0  130.0    3504   
1          buick skylark 320  15.0          8         350.0  165.0    3693   
2         plymouth satellite  18.0          8         318.0  150.0    3436   
3              amc rebel sst  16.0          8         304.0  150.0    3433   
4                ford torino  17.0          8         302.0  140.0    3449   

   acceleration  model_year country  
0          12.0          70     usa  
1          11.5          70     usa  
2          11.0          70     usa  
3          12.0          70     usa  
4          10.5          70     usa

# Well Done!

------------------------------------------------------------

# Hints (Spoiler!)

18. mpg

19. describe() method

20. max() method

21. value_counts() method

22. normalize parameter

23. sort_values() method

24. ascending & inplace parameter

25. sort_index() method; inplace parameter

26. 235.21/mpg

27. describe() method

28. cars.---.---()

29. describe() method

30. unique() method

31. value_counts() method

32. index attribute

33. set_index() method; inplace parameter

34. call is_unique attribute on index

35. call value_counts() method on index

36. cars.index.---  =  "car_model"

37. reset_index() method; inplace parameter

38. columns attribute

39. rename() method; columns parameter